### PR TITLE
WIP defer new conversation call until you post a message

### PR DESCRIPTION
**Pull request by chrisnojima**
Before we would call newConversation whenever you click 'start a chat' from search. This creates a bunch of empty conversations we have to deal with downstream. Instead of doing that we defer making this rpc until you actually post a message. 

In order to do this we make a fake conversation (with fake conversationID) so it will exist in the inbox.

Then when you post we see if we have a fake conversation and then call newConversation, then call post again when its all ready

I also (in the first commit) made the idle callback stuff controllable from local-debug flags to help debug some timing issues

### CODE REVIEW

**Comment by keybase-ci-visdiff** [comment]
The commits 20b01774c618...8be86c73865e introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Menubar-Private_Async_Writing** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1485376396888-20b01774c618-8be86c73865e-linux/Menubar-Private_Async_Writing-20b01774c618.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1485376396888-20b01774c618-8be86c73865e-linux/Menubar-Private_Async_Writing-8be86c73865e.png)  
   Menubar-Private_Async_Writing comparison
</summary>

File: shared/actions/chat.js [hunk 1]
```diff
@@ -169,8 +169,12 @@ const _selectedSelector = (state: TypedState) => {
   return selected
 }
 
+const _inboxSelector = (state: TypedState) => {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: hoist this out

File: shared/actions/chat.js [hunk 2]
```diff
@@ -314,16 +318,16 @@ function _inboxToConversations (inbox: GetInboxLocalRes, author: ?string, follow
   }).filter(Boolean))
 }
 
-function * _clientHeader (messageType: ChatTypes.MessageType, conversationIDKey): Generator<any, ?ChatTypes.MessageClientHeader, any> {
+function * _clientHeader (messageType: ChatTypes.MessageType, conversationIDKey, existingInfo: ?ConversationInfoLocal): Generator<any, ?ChatTypes.MessageClientHeader, any> {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: optionally pass this in if we have it

File: shared/actions/chat.js [hunk 3]
```diff
@@ -391,9 +395,44 @@ function * _getPostingIdentifyBehavior (conversationIDKey: ConversationIDKey) {
   return TlfKeysTLFIdentifyBehavior.chatGuiStrict
 }
 
+function * _startConversationAndPost (action: PostMessage): SagaGenerator<any, any> {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: most of the logic of this new behavior

File: shared/actions/chat.js [hunk 4]
```diff
@@ -391,9 +395,44 @@ function * _getPostingIdentifyBehavior (conversationIDKey: ConversationIDKey) {
...
+  }
+}
+
 function * _postMessage (action: PostMessage): SagaGenerator<any, any> {
   const {conversationIDKey} = action.payload
-  const clientHeader = yield call(_clientHeader, CommonMessageType.text, conversationIDKey)
+  const inbox = yield select(_selectedInboxSelector, conversationIDKey)
+
+  // We created a temp fake conversation? Let's make it and then post
+  if (inbox && inbox.get('isFake')) {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: if we're posting to a fake conversation start one first

File: shared/actions/chat.js [hunk 5]
```diff
@@ -724,10 +763,10 @@ function * _loadInbox (action: LoadInbox): SagaGenerator<any, any> {
     if (incoming.chatInboxConversation) {
       incoming.chatInboxConversation.response.result()
       let conversation: ?InboxState = _inboxConversationToConversation(incoming.chatInboxConversation.params.conv, author, following || {}, metaData)
-      if (conversation && action.payload.newConversationIDKey) {
-        conversation = conversation.set('youCreated', true)
-      }
-      if (conversation && (!conversation.get('isEmpty') || conversation.get('youCreated'))) {
+      if (conversation) {
+        if (action.payload.newConversationIDKey) {
+          conversation = conversation.set('showEvenIfEmpty', true)
```
Comment (chrisnojima) [inline shared/actions/chat.js]: this basically replaces the youCreated to be more generic

File: shared/chat/conversations-list/container.js [hunk 6]
```diff
@@ -19,7 +19,7 @@ class ConversationListContainer extends Component {
 
 export default connect(
   (state: TypedState, {routeSelected}) => ({
-    inbox: state.chat.get('inbox'),
+    inbox: state.chat.get('inbox').filter(i => !i.get('isEmpty') || i.get('showEvenIfEmpty')),
```
Comment (chrisnojima) [inline shared/chat/conversations-list/container.js]: filter out the things we want to hide at this level

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready, please play around with existing and new conversations

File: shared/constants/chat.js [hunk 7]
```diff
@@ -20,6 +20,7 @@ export type AttachmentType = 'Image' | 'Other'
 
 export type ConversationID = RPCConversationID
 export type ConversationIDKey = string
+export type FakeConversationIDKey = string
```
Comment (gabriel) [inline shared/constants/chat.js]: "Fake" is a little scary to me, maybe should be "Temp"?

Comment (chrisnojima) [inline shared/constants/chat.js]: i'll make this switch

**Comment by chrisnojima** [comment]
- [x] Doens't work if the inbox reloads before we get to it, fixing that now

**Comment by chrisnojima** [comment]
fixing this requires me to refactor our store a bunch, deferring until that's done

**Pull request closed without merging**